### PR TITLE
Ensure tests representing containing types of suites are synthesized if necessary

### PR DESCRIPTION
**Pull request by stmontgomery**
Ensure that when a suite is nested in one or more containing suite types, and those outer suites do not have an explicit `@Suite` attribute, we synthesize `Test` instances representing all containing suites.

### Checklist:

- [x] Code and documentation should follow the style of the [Style Guide](https://github.com/apple/swift-testing/blob/main/Documentation/StyleGuide.md).
- [x] If public symbols are renamed or modified, DocC references should be updated.

Resolves rdar://127770218

### CODE REVIEW

File: Sources/Testing/Parameterization/TypeInfo.swift [hunk 1]
```diff
@@ -276,11 +276,18 @@ extension TypeInfo {
     }
     return nil
   }
+}
 
-  /// A sequence of instances of this type representing the types that
-  /// recursively contain it, starting with the immediate parent (if any.)
-  var allContainingTypeInfo: some Sequence<Self> {
-    sequence(first: self, next: \.containingTypeInfo).dropFirst()
+extension Test {
```
Comment (grynspan) [inline Sources/Testing/Parameterization/TypeInfo.swift:281]: Just cleanup?

**Comment by stmontgomery** [comment]
@swift-ci please test

**Comment by stmontgomery** [comment]
@swift-ci please test Linux

File: Tests/TestingTests/RunnerTests.swift [hunk 2]
```diff
@@ -501,7 +501,7 @@ final class RunnerTests: XCTestCase {
 
   func testPoundIfTrueTestFunctionRuns() async throws {
     let testStarted = expectation(description: "Test started")
-    testStarted.expectedFulfillmentCount = 4
+    testStarted.expectedFulfillmentCount = 5
```
Comment (briancroom) [inline Tests/TestingTests/RunnerTests.swift:504]: These increments are because a synthesized suite for `RunnerTests` is now included in the runner plan?

File: Tests/TestingTests/PlanTests.swift [hunk 3]
```diff
@@ -421,6 +421,15 @@ struct PlanTests {
     #expect(plan.stepGraph.subgraph(at: typeInfo.fullyQualifiedNameComponents + CollectionOfOne("reserved1(reserved2:)")) != nil)
   }
 
+  @Test("Containing suite types without @Suite are synthesized")
+  func containingSuiteSynthesis() async {
+    let plan = await Runner.Plan(selecting: OutermostSuite.self)
+    let testNames = plan.steps.map(\.test.name)
+    #expect(testNames.contains("OutermostSuite"))
```
Comment (briancroom) [inline Tests/TestingTests/PlanTests.swift:428]: Is `PlanTests` part of this list also?

File: Sources/Testing/Test+Discovery.swift [hunk 4]
```diff
@@ -80,32 +80,23 @@ extension Test {
...
+  private static func _synthesizeSuiteTypes(into tests: inout [ID: Self]) {
+    // We won't have an accurate source location for synthesized suites, so use
+    // a "runtime inferred" sentinel value instead.
+    //
+    // FIXME: Consider either changing `Test.sourceLocation` to Optional (an
+    // API-breaking change) or modifying Runner.Plan so that it doesn't require
+    // an actual Test instance to represent synthesized parent suites.
+    lazy var inferredSourceLocation = SourceLocation(fileID: "<runtime-inferred>", filePath: "<runtime-inferred>", line: 0, column: 0)
+
+    for test in tests.values {
```
Comment (briancroom) [inline Sources/Testing/Test+Discovery.swift:92]: What do the performance characteristics of this look like? Do we possibly need more caching to keep the repeated lookups from becoming a bottleneck?

File: Tests/TestingTests/RunnerTests.swift (hunk 2 above)
Comment (stmontgomery) [reply Tests/TestingTests/RunnerTests.swift:504]: Yes, exactly. Many of the fixture suites in this file are nested right within the `RunnerTests` XCTestCase subclass.

File: Tests/TestingTests/PlanTests.swift (hunk 3 above)
Comment (stmontgomery) [reply Tests/TestingTests/PlanTests.swift:428]: Yes. I nested the fixture suite within `PlanTests` right before posting the PR and forgot to come back and add it here for clarity. It isn't strictly necessary to validate that the fix works (checking for the presence of `OutermostSuite` is sufficient). But including `PlanTests` would be better, so I'll add that and also restructure the `#expect` so that it requires a strict order for the test names.

**Comment by stmontgomery** [comment]
@swift-ci please test

File: Sources/Testing/Test+Discovery.swift (hunk 4 above)
Comment (stmontgomery) [reply Sources/Testing/Test+Discovery.swift]: I don't think the performance characteristics of this function are changed in a significant way. However, I have never really loved the approach that this method uses, which involves mapping the tests into a dictionary, iterating every test and all of its ancestors, constructing a test ID for each one, and testing whether it's already in the dictionary.

I decided to take a different approach with this PR, which requires broader changes but I think has more benefits overall. In my latest commit, we no longer construct this temporary dictionary, and instead synthesize suites during planning while traversing the test graph. While implementing and debugging this, I also noticed some opportunities to simplify and optimize the test graph by pruning nodes for tests which have been filtered out.

**Comment by stmontgomery** [comment]
@swift-ci please test

File: Sources/Testing/Test+Discovery.swift (hunk 4 above)
Comment (grynspan) [reply Sources/Testing/Test+Discovery.swift]: The dictionary (or a set) is necessary to avoid duplicate test or suite records which can potentially appear multiple times in the raw data we're walking. Does your new code dedup the data?

Comment (stmontgomery) [reply Sources/Testing/Test+Discovery.swift]: Yeah, they are de-duplicated by their placement in the test graph (which relies on a dictionary)

**Comment by stmontgomery** [comment]
@swift-ci please test

**Comment by stmontgomery** [comment]
@swift-ci please test

File: Sources/Testing/Running/Configuration.TestFilter.swift [hunk 5]
```diff
@@ -422,3 +485,60 @@ extension Configuration.TestFilter {
     self = combining(with: other, using: op)
   }
 }
+
+// MARK: - Filterable types
+
+extension Configuration.TestFilter.Kind {
```
Comment (grynspan) [inline Sources/Testing/Running/Configuration.TestFilter.swift:491]: Are the changes in this file actually related to suite synthesis?

File: Sources/Testing/Running/Runner.Plan.swift [hunk 6]
```diff
@@ -304,6 +344,20 @@ extension Runner.Plan {
...
+  ///
+  /// When a suite is synthesized, it does not have accurate source location, so
+  /// the value of this property is a "runtime inferred" source location which
+  /// can be used instead.
+  //
+  // FIXME: Consider either changing `Test.sourceLocation` to Optional (an
+  // API-breaking change) or modifying Runner.Plan so that it doesn't require
+  // an actual Test instance to represent synthesized parent suites, thereby
+  // avoiding the need for a source location for them.
+  fileprivate static let synthesized = SourceLocation(fileID: "<runtime-inferred>", filePath: "<runtime-inferred>", line: 0, column: 0)
```
Comment (grynspan) [inline Sources/Testing/Running/Runner.Plan.swift:358]: The malformed file ID will cause crashes.
Line 0 and column 0 aren't allowed and will also cause crashes. Use 1 for each.

"<runtime-inferred>" reads to me like the Swift runtime inferred them, not that they were synthesized at run time. Bikeshed a bit?

File: Sources/Testing/Test.swift [hunk 7]
```diff
@@ -192,29 +192,18 @@ public struct Test: Sendable {
...
   }
 
-  /// Whether or not this instance was synthesized at runtime.
-  ///
-  /// During test planning, suites that are not explicitly marked with the
-  /// `@Suite` attribute are synthesized from available type information before
-  /// being added to the plan. For such suites, the value of this property is
-  /// `true`.
-  @_spi(ForToolsIntegrationOnly)
-  public var isSynthesized = false
```
Comment (grynspan) [inline Sources/Testing/Test.swift:202]: This property is useful for diagnosing issues with synthesized instances—why did you remove it?

**Review by grynspan — CHANGES_REQUESTED** [review]
See comments above.

File: Sources/Testing/Running/Runner.Plan.swift (hunk 6 above)
Comment (stmontgomery) [reply Sources/Testing/Running/Runner.Plan.swift:358]: > Line 0 and column 0 aren't allowed and will also cause crashes. Use 1 for each.

Hmm… why is this not crashing currently if that's the case? This property is being accessed.

> `"<runtime-inferred>"` reads to me like the Swift runtime inferred them, not that they were synthesized at run time. Bikeshed a bit?

The compiler sometimes generates code with locations that look like `<compiler-generated>:0`, and this is what inspired me to use the `<...>` notation. Any alternative wording suggestions? `"<inferred>"` ? 

I also want to draw attention to the "FIXME" comment above this, which suggests future ways we could avoid needing this property altogether.

File: Sources/Testing/Test.swift (hunk 7 above)
Comment (stmontgomery) [reply Sources/Testing/Test.swift:202]: It doesn't appear to be used anywhere

File: Sources/Testing/Running/Runner.Plan.swift (hunk 6 above)
Comment (stmontgomery) [reply Sources/Testing/Running/Runner.Plan.swift:358]: > Hmm… why is this not crashing currently if that's the case? This property is being accessed.

Ah, right. `didSet` isn't used for property assignments in initializers

File: Sources/Testing/Running/Configuration.TestFilter.swift (hunk 5 above)
Comment (stmontgomery) [reply Sources/Testing/Running/Configuration.TestFilter.swift:491]: I've expanded the scope of this PR to more than just the bug fix. I'm going to rewrite the PR description to summarize, although if you think it would help, I could tease this apart into separate PRs, and just use this one to focus on the original bug fix (related to test suite synthesis).

File: Sources/Testing/Running/Runner.Plan.swift (hunk 6 above)
Comment (grynspan) [reply Sources/Testing/Running/Runner.Plan.swift]: > Ah, right. didSet isn't used for property assignments in initializers

Indeedy.

> Any alternative wording suggestions? "<inferred>" ?

Off the top of my head:

- `<synthesized>`
- `<compiler-generated>` (i.e. just lie for consistency with other bits of the stack)
- `<library-generated>`

But the string needs to be a valid file ID anyway (file path ideally should be valid too, but that's less urgent.) _This is why I was using the source location of the nearest real test_. It wasn't just arbitrary.

File: Sources/Testing/Test.swift (hunk 7 above)
Comment (grynspan) [reply Sources/Testing/Test.swift:202]: It's not actively used but it's _useful_ when I'm debugging stuff.

File: Sources/Testing/Running/Configuration.TestFilter.swift (hunk 5 above)
Comment (grynspan) [reply Sources/Testing/Running/Configuration.TestFilter.swift:491]: Yes, I do think these unrelated changes belong in their own PR. Unless they really are related, in which case please elaborate.

File: Sources/Testing/Running/Runner.Plan.swift [hunk 8]
```diff
@@ -304,6 +344,20 @@ extension Runner.Plan {
...
+  ///
+  /// When a suite is synthesized, it does not have accurate source location, so
+  /// the value of this property is a "runtime inferred" source location which
+  /// can be used instead.
+  //
+  // FIXME: Consider either changing `Test.sourceLocation` to Optional (an
+  // API-breaking change) or modifying Runner.Plan so that it doesn't require
+  // an actual Test instance to represent synthesized parent suites, thereby
+  // avoiding the need for a source location for them.
+  fileprivate static let synthesized = SourceLocation(fileID: "<unknown-module>/<unknown-file>", filePath: "<unknown-file>", line: 0, column: 0)
```
Comment (grynspan) [inline Sources/Testing/Running/Runner.Plan.swift:358]: Did you check to see if this breaks total ordering for tests now that they may appear to be located in a file with this name?

Also the module is not unknown for these types and should be correctly specified.

Comment (stmontgomery) [reply Sources/Testing/Running/Runner.Plan.swift:358]: > Did you check to see if this breaks total ordering for tests now that they may appear to be located in a file with this name?

Thanks for that context! I'm going to change the portion of this PR related to inferring the SourceLocation to ensure it continues using the location of some child, to ensure the sort ordering of tests remains unchanged. This should address this and your [other PR comment](https://github.com/swiftlang/swift-testing/pull/407#discussion_r1835109103) above.

One nice thing about my new approach in this PR (synthesizing while recursively traversing the graph during planning) is that it can ensure that any synthesized suite always uses a source location of a child test which is a minimum distance from it in the graph, meaning the inferred source location should be closer to accurate more often. I'll also add unit tests for it.

File: Sources/Testing/Running/Runner.Plan.swift (hunk 6 above)
Comment (stmontgomery) [reply Sources/Testing/Running/Runner.Plan.swift]: > > Ah, right. didSet isn't used for property assignments in initializers
> 
> Indeedy.

I spun off a separate PR to improve this: https://github.com/swiftlang/swift-testing/pull/817

File: Sources/Testing/Running/Runner.Plan.swift (hunk 8 above)
Comment (grynspan) [reply Sources/Testing/Running/Runner.Plan.swift:358]: I see. We should definitely fix the module names, but if the filename isn't an issue, then we can leave it stubbed out.

File: Sources/Testing/Test.swift (hunk 7 above)
Comment (stmontgomery) [reply Sources/Testing/Test.swift:202]: What if we made `isSynthesized` into an `internal`-only, non-recursive `SuiteTrait` instead? Then we could eliminate this stored property, while keeping the ability to see whether a suite is synthesized while debugging.

And if we take that approach, I would like to conditionalize the trait's inclusion using `#if DEBUG`.

File: Tests/TestingTests/MiscellaneousTests.swift [hunk 9]
```diff
@@ -552,17 +552,6 @@ struct MiscellaneousTests {
     #expect(id.keyPathRepresentation == [""])
   }
 
-  @Test("Test.all deduping")
```
Comment (grynspan) [inline Tests/TestingTests/MiscellaneousTests.swift:555]: Why did you remove this test? Did it start failing? Can it be refactored?

Comment (stmontgomery) [reply Tests/TestingTests/MiscellaneousTests.swift:555]: This PR removes `Test.testsByID()` and modifies `Test.all` to directly return all runtime-enumerated test instances, so I didn't see a remaining need for this test.

Comment (grynspan) [reply Tests/TestingTests/MiscellaneousTests.swift:555]: We still want to test that content is deduped because the raw content we collect can have duplicates.

File: Sources/Testing/Running/Configuration.TestFilter.swift (hunk 5 above)
Comment (stmontgomery) [reply Sources/Testing/Running/Configuration.TestFilter.swift:491]: I've split this out into #820

File: Tests/TestingTests/MiscellaneousTests.swift (hunk 9 above)
Comment (stmontgomery) [reply Tests/TestingTests/MiscellaneousTests.swift:555]: Ok, I think that would be better to test at the end of the planning process, whereas this test is validating an intermediary stage (after runtime discovery but before planning). I'll refactor it.

Comment (grynspan) [reply Tests/TestingTests/MiscellaneousTests.swift:555]: Deduping can be a natural consequence of us building the initial test graph, FWIW.

Comment (stmontgomery) [reply Tests/TestingTests/MiscellaneousTests.swift:555]: Oh yeah, and I recall now that's actually why I didn't bother replacing or refactoring this test when I removed it. I think we're OK as-is

File: Sources/Testing/Running/Configuration.TestFilter.swift [hunk 10]
```diff
@@ -191,7 +191,7 @@ extension Configuration.TestFilter {
 extension Configuration.TestFilter {
   /// An enumeration which represents filtering logic to be applied to a test
   /// graph.
-  fileprivate enum Operation: Sendable {
```
Comment (grynspan) [inline Sources/Testing/Running/Configuration.TestFilter.swift:194]: We don't _usually_ bother naming generic parameters more than `T`, `U`, etc.… should we start?

File: Sources/Testing/Running/Runner.Plan.swift [hunk 11]
```diff
@@ -154,6 +154,52 @@ extension Runner.Plan {
...
   }
 
+  /// Recursively synthesize test instances representing suites for all missing
+  /// values in the specified test graph.
+  ///
+  /// - Parameters:
+  ///   - graph: The graph in which suites should be synthesized.
+  ///   - nameComponents: The name components of the suite to synthesize, based
+  ///     on the key path from the root node of the test graph to `graph`.
+  private static func _recursivelySynthesizeSuites(in graph: inout Graph<String, Test?>, nameComponents: [String] = []) {
```
Comment (grynspan) [inline Sources/Testing/Running/Runner.Plan.swift:164]: Test coverage of this function at 100%, yeah?

Comment (stmontgomery) [reply Sources/Testing/Running/Runner.Plan.swift:164]: Yes!

File: Sources/Testing/Running/Configuration.TestFilter.swift (hunk 10 above)
Comment (stmontgomery) [reply Sources/Testing/Running/Configuration.TestFilter.swift:194]: I can switch it to `T` for now, I just found this one instance a bit confusing out of context

**Comment by stmontgomery** [comment]
@swift-ci please test

File: Sources/Testing/Test.swift (hunk 7 above)
Comment (stmontgomery) [reply Sources/Testing/Test.swift]: For this PR, I have restored the `isSynthesized` property, and actually adopted it in my new unit tests so it's no longer unused

**Comment by stmontgomery** [comment]
@swift-ci please test
